include/triton/Tools/LinearLayout.h

```diff
@@ -0,0 +1,433 @@
+//                          | 1 |
+//                          | 1 |
+//
+// This works, but it's cumbersome.  It's more compact to think of the vector
+// `a` as an M-bit unsigned integer, and each column Bi of the matrix B as an
+// N-bit integer.  Here's the same matrix-vector product written this way.
+//
+//   = | 1 2 14 12 | × 6
+//   = | 1 2 14 12 | × 0b0110
+//   = (1 × 0) ⊕ (2 × 1) ⊕ (14 × 1) ⊕ (12 × 0)
+//   = 2 ⊕ 14
+//   = 12.
+//
+// And we confirm that our answer of 12 is equal to the binary value 0b1100 we
+// got before.
+//
+// Notice that the function F(a) is fully specified by the matrix B, and that
+// the four columns of B tell us the values of F at power-of-two values for `a`,
+// namely F(1), F(2), F(4), and F(8).  In other words, we specify four values of
+// F (we call these the function's "basis vectors" or its "bases") and we can
+// then compute any other value by xor'ing together subsets of the bases.
+//
+// In the case of a 1D -> 1D layout, the implementation of an LL is
+// straightforward from the mathematical description.  If the LL is
+// higher-dimensional, we can "stack" the bit vectors to create 1D vectors.  For
+// example, if we have a 2D LL and we're given input tuple (0b0011, 0b1100), we
+// can treat this like a 1D input 0b0011'1100 and then do the regular 1D LL
+// computation.  Similarly we can "unstack" the output from 1D to ND.
+//
+// With stacking and unstacking, the linearity of our function in GF(2) directly
+// implies the "linearity rule" from earlier.  And that's all we need in order
+// to define linear layouts mathematically!
+//
+// # Comaprison to Nvidia CuTe
+//
+// (Note, I'm not an expert on CuTe; this is my best understanding.)
+//
+// CuTe is a programmatic layout system that's part of Nvidia CUTLASS; see
+// https://github.com/NVIDIA/cutlass/blob/629f465/media/docs/cute/00_quickstart.md
+//
+// LLs and CuTe solve similar problems.  Before CuTe, CUTLASS v2 had many
+// handcrafted layouts, "RowMajor", "VoltaTensorOpMultiplicandCongruous", etc,
+// see https://www.youtube.com/watch?v=QLdUML5MCfE&t=574s.  Each of these was a
+// special case.  CUTLASS v3 introduced CuTe layouts, which are a programmable
+// layout that subsumes all of these special cases.  The CUTLASS folks say this
+// simplified CUTLASS, in the same way that we hope LLs will simplify Triton.
+//
+// Like CuTe layouts, LLs are also flexible and composible.  But there are also
+// some key differences.
+//
+//  - CuTe layouts are MD -> 1D, whereas LLs are MD -> ND.  This is only a UI
+//    difference, because an MD -> 1D layout can be "unstacked" into MD -> ND.
+//  - Dimensions in LLs are named; CuTe dimensions are numbered.
+//  - CuTe layouts can be nested; LLs cannot be.  (Nesting doesn't give CuTe
+//    layouts additional power; any nested layout can be flattened.)
+//  - CuTe layouts support non-power-of-two shapes; LLs do not.  In particular
+//    this means that LLs cannot represent padded layouts.
```

Comment:
Any particular reason why LLs do not implement non powers of two? How does the obvious strategy of embedding a non-power of two into the next power of two fail?

Reply:
Ah, right, I completely forgot for a minute about the power of two restriction in the triton langaguage, as we do actually have to worry about padding at a PyTorch level. ooops